File: ruby/anagram/anagram.rb
class Anagram
  def initialize(word)
    @word = word.downcase
  end

  def match(words)
    words.select {|word| anagram? word.downcase}
  end

  private

  def anagram?(word)
    return false if word == @word
    word.chars.sort == @word.chars.sort
  end

end

File: elixir/anagram/lib/anagram.ex
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t(), [String.t()]) :: [String.t()]
  def match(base, candidates) do
    anagram = base |> rearrange

    for word <- candidates,
        word !== base,
        String.downcase(word) !== String.downcase(base),
        anagram === rearrange(word),
        into: [],
        do: word
  end

  defp rearrange(word) do
    word |> String.downcase() |> Kernel.to_charlist() |> Enum.sort()
  end
end
